src/Handler/Caching.hs

```diff
@@ -23,13 +26,27 @@ import Data.Aeson (encode)
 
 import Handler.Utils
 
+data OkToCache
+  = OkToCache
+  | NotOkToCache
+  deriving (Show, Eq, Ord)
+
 -- | This function allows an upstream server (such as nginx) to cache a
 -- response, by writing the response body to a file before sending it to the
 -- client. Afterwards, the upstream server can respond to requests for the same
 -- path without having to proxy to the Yesod application, and any caching
 -- mechanisms that the upstream server supports can additionally be used.
-cache :: (a -> LB.ByteString) -> String -> Handler a -> Handler a
-cache toLbs basename action = action >>= (\body -> write body $> body)
+--
+-- We use Yesod's short-circuiting behaviour here to ensure that this code is
```

Comment:
If the `Handler` action that gets passed in uses `sendResponseStatus`, that short-circuits and sends a response immediately, and no further `Handler` code is run - in particular, this code shouldn't be reached.